### DISCUSSION TITLE
Reorder class constants/properties/methods

### DISCUSSION THREAD

**Discussion by cviebrock** [phpactor/phpactor#2692] [Ideas]
I like the [Rearrange Code](https://www.jetbrains.com/help/phpstorm/reformat-and-rearrange-code.html) feature in PhpStorm that will take a class and sort it's constants, properties, and methods according to a set of rules you can define.  e.g.:

1. constants, grouped by visibility, sorted by name
2. properties, grouped by visibility, sorted by name
3. the `__construct` method, if it exists
4. other class methods, grouped by visibility, sorted by name
5. magic methods, sorted by name

I was hoping to add something like this as a CLI command for our codebase, so that we're not reliant on all devs using PhpStorm.

Does phpactor support this currently?  Any plans to?  Or are there other tools that might help here?

Thanks!

**Comment by mamazu** [comment]
Hey thanks for bringing this idea up. Currently phpactor doesn't have a feature like this. For example When adding new properties to a class phpactor will put them before methods but it won't fix the entire class. As far as I know there is no such feature planned but it shouldn't be too hard to implement just moving some parts around.

If you want to have other tool recommendations though. [ECS](https://github.com/easy-coding-standard/easy-coding-standard) would be my tool of choice and this is the rule your looking for [orders class elements](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/rules/class_notation/ordered_class_elements.rst). If you're already using php-cs you can just use the rule.

Sidenote: I'm curious if you order methods by name and you rename a method and it might need to be moved - wouldn't that make diffs really hard to read?

**Comment by dantleech** [comment]
php-cs-fixer has a native integration with Phpactor and has the rule as Mamazu mentioned. I have no interest in implemting it as it'n one of 1000 fixes that a cs fixer can do already :)
